### PR TITLE
Ensure that the Assertion is signed when doing signature checks

### PR DESCRIPTION
**Pull request by jefff**
Under some circumstances, it could be possible that something else in
the SAML Response is signed instead of the Assertion. Double check that
the Assertion itself is part of the signature before considering the
Response valid.

### CODE REVIEW

File: lib/saml2.coffee [hunk 1]
```diff
@@ -340,10 +365,23 @@ parse_authn_response = (saml_response, sp_private_key, idp_certificates, allow_u
         cb_wf null, assertion[0].toString()
     (result, cb_wf) ->
       debug result
-      decrypted_assertion = (new xmldom.DOMParser()).parseFromString(result)
-      unless ignore_signature || _.some(idp_certificates, (cert) -> check_saml_signature result, cert)
-        return cb_wf new Error("SAML Assertion signature check failed! (checked #{idp_certificates.length} certificate(s))")
-      cb_wf null
+      if ignore_signature
+        decrypted_assertion = (new xmldom.DOMParser()).parseFromString(result)
```
Comment (alsmola) [inline lib/saml2.coffee]: What's this line doing?

File: lib/saml2.coffee [hunk 2]
```diff
@@ -161,11 +161,36 @@ check_saml_signature = (xml, certificate, cb) ->
...
+    uri = ref.uri
+    if uri[0] is '#'
+      uri = uri.substring(1)
+
+    elem = []
+    if uri is ""
+      elem = xmlcrypto.xpath(doc, "//*")
+    else
+      for idAttribute in ["Id", "ID"]
+        elem = xmlcrypto.xpath(doc, "//*[@*[local-name(.)='" + idAttribute + "']='" + uri + "']")
```
Comment (alsmola) [inline lib/saml2.coffee]: This makes me a little nervous about [XPath Injection] (https://www.owasp.org/index.php/XPATH_Injection). Should we escape single quotes here?

File: test/saml2.coffee [hunk 3]
```diff
@@ -104,13 +104,14 @@ describe 'saml2', ->
...
       it 'accepts signed xml', ->
-        assert saml2.check_saml_signature(get_test_file("good_assertion.xml"), get_test_file("test.crt"))
+        result = saml2.check_saml_signature(get_test_file("good_assertion.xml"), get_test_file("test.crt"))
+        assert.deepEqual result, [get_test_file("good_assertion.xml")]
 
       it 'rejects xml without a signature', ->
-        assert.equal false, saml2.check_saml_signature(get_test_file("unsigned_assertion.xml"), get_test_file("test.crt"))
+        assert.equal null, saml2.check_saml_signature(get_test_file("unsigned_assertion.xml"), get_test_file("test.crt"))
 
       it 'rejects xml with an invalid signature', ->
```
Comment (alsmola) [inline test/saml2.coffee]: Worth adding a test that would fail with the previous code, and pass with this one, to help prevent regressions?

Comment (jefff) [inline test/saml2.coffee]: Good idea. Test added. 510d0a0

**Pull request merged**